src/api/datastores.rs

```diff
@@ -0,0 +1,323 @@
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+
+use futures::future::Either;
+use futures::stream::Stream;
+use futures::{future, Future};
+use hyper::{header, Body, Chunk, Request, Response};
+
+// This file is part of MinSQL
+// Copyright (c) 2019 MinIO, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+use crate::api::{ListResponse, SafeOutput, ViewSet};
+use crate::config::{Config, DataStore};
+use crate::http::{return_400, return_404, return_500, ResponseFuture};
+use crate::storage::{delete_object_metabucket, put_object_metabucket};
+
+pub struct ApiDataStores {
+    config: Arc<RwLock<Config>>,
+}
+
+impl SafeOutput for DataStore {
+    fn safe(&mut self) {
+        self.secret_key = "*********".to_string();
+    }
+}
+
+impl ApiDataStores {
+    pub fn new(cfg: Arc<RwLock<Config>>) -> ApiDataStores {
+        ApiDataStores { config: cfg }
+    }
+
+    // Parses the datastore from the create body; returns error response in
+    // case it is not valid.
+    fn parse_create_body(
+        entire_body: Vec<u8>,
+        cfg: Arc<RwLock<Config>>,
+    ) -> Result<DataStore, Response<Body>> {
+        let payload: String = match String::from_utf8(entire_body.to_vec()) {
```

Comment:
Any reason to not use the `?` style? We were already using it before and it looks like we are going back in terms of code improvements.